src/main/java/org/ccci/gto/persistence/DeadLockRetryAspectSupport.java

```diff
@@ -0,0 +1,101 @@
+package org.ccci.gto.persistence;
+
+import static org.ccci.gto.persistence.DeadLockRetry.DEFAULT_ATTEMPTS;
+
+import org.ccci.gto.persistence.tx.Closure;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.core.Ordered;
+import org.springframework.dao.ConcurrencyFailureException;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.persistence.EntityManager;
+import javax.persistence.OptimisticLockException;
+import javax.validation.constraints.NotNull;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+public abstract class DeadLockRetryAspectSupport implements Ordered {
+    private static final Logger LOG = LoggerFactory.getLogger(DeadLockRetryAspectSupport.class);
+
+    @NotNull
+    private EntityManagerResolver entityManagerResolver;
+
+    @Nonnull
+    private List<DeadLockDetector> deadLockDetectors = new ArrayList<>();
+
+    private int defaultAttempts = 3;
+
+    private int order = LOWEST_PRECEDENCE;
+
+    @Autowired(required = false)
+    public void setDeadLockDetectors(@Nullable final List<DeadLockDetector> detectors) {
+        deadLockDetectors = detectors != null ? detectors : new ArrayList<>();
+    }
+
+    @Autowired
+    public void setEntityManagerResolver(@Nonnull final EntityManagerResolver resolver) {
+        entityManagerResolver = resolver;
+    }
+
+    public void setDefaultAttempts(final int attempts) {
+        defaultAttempts = attempts;
+    }
+
+    @Override
+    public int getOrder() {
+        return order;
+    }
+
+    /**
+     * Sets the order.
+     *
+     * @param order the order to set
+     */
+    public void setOrder(final int order) {
+        this.order = order;
+    }
+
+    @Nonnull
+    private Optional<EntityManager> resolveEntityManager(@Nonnull final String unitName) {
+        return "".equals(unitName) ? entityManagerResolver.defaultEntityManager() :
+                entityManagerResolver.resolveEntityManager(unitName);
+    }
+
+    protected final <X extends Throwable> Object wrapJoinPoint(@Nonnull final Closure<Object, X> joinPoint,
```

Comment:
I'm a little late, but a better name would have been something like 'executeWithDeadlockRetry' or 'executeRetryingDeadlocks'. The word 'wrap' communicates to me that this is returning a join point, not executing it immediately.

Reply:
just saw this, will update it for the next release :)